File: openframeworks/drawing/src/stroke.cpp
#include "stroke.h"

stroke :: stroke() {
    strokeSize = 10;
    strokeColor = ofColor(0);
    smoothReps = 10;
    splitReps = 2;
}

stroke :: stroke(ofColor c) {
    strokeColor = c;
    strokeSize = 10;
    smoothReps = 10;
    splitReps = 2;
}

stroke :: stroke(float s) {
    strokeColor = ofColor(0);
    strokeSize = s;
    smoothReps = 10;
    splitReps = 2;
}

stroke :: stroke(ofColor c, float s) {
    strokeColor = c;
    strokeSize = s;
    smoothReps = 10;
    splitReps = 2;
}

void stroke :: update() {
    //
}

void stroke :: draw() {
    ofNoFill();
    ofBeginShape();
    for (int i=0; i<points.size(); i++) {
        ofSetLineWidth(strokeSize);
        ofSetColor(strokeColor);
        ofVertex(points[i] -> x, points[i] -> y, points[i] -> z);
    }
    ofEndShape();
}

void stroke :: run() {
    update();
    draw();
}

void stroke :: splitStroke() {
    for (int i = 1; i < points.size(); i+=2) {
        ofVec3f center = *points[i];
        ofVec3f lower = *points[i-1];
        float x = (center.x + lower.x) / 2;
        float y = (center.y + lower.y) / 2;
        float z = (center.z + lower.z) / 2;
        ofVec3f p = ofVec3f(x, y, z);
        //points.insert(points.begin() + i, &p);
    }
}

void stroke :: smoothStroke() {
    float weight = 18;
    float scale = 1.0 / (weight + 2);
    int nPointsMinusTwo = points.size() - 2;
    ofVec3f lower, upper, center;
    
    for (int i = 1; i < nPointsMinusTwo; i++) {
        lower = *points[i-1];
        center = *points[i];
        upper = *points[i+1];
        
        center.x = (lower.x + weight * center.x + upper.x) * scale;
        center.y = (lower.y + weight * center.y + upper.y) * scale;
        center.z = (lower.z + weight * center.z + upper.z) * scale;
        *points[i] = center;
    }
}

void stroke :: refine() {
    for (int i=0; i<splitReps; i++) {
        splitStroke();
        smoothStroke();
    }
    for (int i=0; i<smoothReps - splitReps; i++) {
        smoothStroke();
    }
}

File: processingpy/drawing/Stroke.py
class Stroke(object):

  def __init__(self, c=color(0), s=10):
    self.points = []
    self.strokeSize = s
    self.smoothReps = 10
    self.splitReps = 2
    self.strokeColor = c
   
  def update(self):
    pass

  def draw(self):
    noFill()
    beginShape()
    for pt in self.points:
      strokeWeight(self.strokeSize)
      stroke(self.strokeColor)
      vertex(pt.x, pt.y, pt.z)
    
    endShape()

  def run(self):
    self.update()
    self.draw()

  def smoothStroke(self):
        weight = 18
        scale = 1.0 / (weight + 2)
        nPointsMinusTwo = len(self.points) - 2
        lower = 0
        upper = 0
        center = 0

        for i in range(1, nPointsMinusTwo):
            lower = self.points[i-1]
            center = self.points[i]
            upper = self.points[i+1]

            center.x = (lower.x + weight * center.x + upper.x) * scale
            center.y = (lower.y + weight * center.y + upper.y) * scale
        
  def splitStroke(self):
    for i in range(1, len(self.points), 2):
      center = self.points[i]
      lower = self.points[i-1]
      x = (center.x + lower.x) / 2
      y = (center.y + lower.y) / 2
      z = (center.z + lower.z) / 2
      p = PVector(x, y, z)
      self.points.insert(i, p)

  def refine(self):
    for i in range(0, self.splitReps):
      self.splitStroke()  
      self.smoothStroke()  
    
    for i in range(0, self.smoothReps - self.splitReps):
      self.smoothStroke()    
     
  
  
